### PR TITLE
Feature comment target

### PR DESCRIPTION
**Pull request by tasdomas**
This is a WIP branch

- [x] #1202
- [x] #1241
- [x] test for no breaking changes
- [x] open documentation follow-up https://github.com/iterative/cml.dev/pull/415

### CODE REVIEW

File: src/drivers/bitbucket_cloud.js
Comment (DavidGOrtega) [inline src/drivers/bitbucket_cloud.js:57]:
```suggestion
  async issueCommentUpsert(opts = {}) {
    const { projectPath } = this;
    const { issueId, report, id } = opts;
    const endpoint = `/repositories/${projectPath}/issues/${issueId}/comments/${id ? `${id}/` : ''}`;
    return (
      await this.request({
        endpoint,
        method: id ? 'PUT' : 'POST',
        body: JSON.stringify({ content: { raw: report } })
      })
    ).links.html.href;
  }

  async issueCommentCreate(opts = {}) {
    const { id, ...rest } = opts;
    return this.issueCommentUpsert(rest)
  }
  
  async issueCommentUpdate(opts = {}) {
    if (!opts.id) throw new Error('Id is missing updating comment')
    return this.issueCommentUpsert(opts)
  }
```

File: src/drivers/gitlab.js
Comment (DavidGOrtega) [inline src/drivers/gitlab.js:364]:
```suggestion
  async issueCommentUpsert(opts = {}) {
    const projectPath = await this.projectPath();
    const { issueId, report, id: commentId } = opts;

    const endpoint = `/projects/${projectPath}/issues/${issueId}/notes${ commentId ? `/${commentId}` : ''}`;
    const body = new URLSearchParams();
    body.append('body', report);

    const { id } = await this.request({
      endpoint,
      method: commentId  ? 'PUT' : 'POST',
      body
    });

    return `${this.repo}/-/issues/${issueId}#note_${id}`;
  }

  async issueCommentCreate(opts = {}) {
    const { id, ...rest } = opts;
    return this.issueCommentUpsert(rest)
  }
  
  async issueCommentUpdate(opts = {}) {
    if (!opts.id) throw new Error('Id is missing updating comment')
    return this.issueCommentUpsert(opts)
  }
```

File: src/drivers/github.js
Comment (DavidGOrtega) [inline src/drivers/github.js:560]:
```suggestion
    const { id, report } = opts;
    if (!id) throw new Error('Id is missing updating comment')
```

**Review by DavidGOrtega — DISMISSED** [review]
Please consider this suggestion just to reduce some boilerplate.
If you still want to continue with the current code is still OK
🙏

**Comment by karl-richter** [comment]
Hi @tasdomas, do I understand it correct that this feature would allow to comment (and update comments) on a pull request given the respective pr number on the repo? I was about to open an issue as I seem to have run into a deadlock with commenting on a pr in the target repo when opening the pr from a branch in a fork. Seems like this could resolve the issue.

**Comment by tasdomas** [comment]
> Hi @tasdomas, do I understand it correct that this feature would allow to comment (and update comments) on a pull request given the respective pr number on the repo? I was about to open an issue as I seem to have run into a deadlock with commenting on a pr in the target repo when working with forks. Seems like this could resolve the issue.

Yes, this will be one of the features this will introduce

**Comment by karl-richter** [comment]
> > Hi @tasdomas, do I understand it correct that this feature would allow to comment (and update comments) on a pull request given the respective pr number on the repo? I was about to open an issue as I seem to have run into a deadlock with commenting on a pr in the target repo when working with forks. Seems like this could resolve the issue.
> 
> Yes, this will be one of the features this will introduce

Sounds awesome! Do you have an eta for the pr / whats your release cycle?
Let me know if I can support with getting this shipped!

**Comment by DavidGOrtega** [comment]
@tasdomas what happened with this effort?

**Comment by tasdomas** [comment]
> @tasdomas what happened with this effort?
This is a WIP at the moment, this branch is a feature branch focused on introducing this change.

Once we've landed #1241 and done all the necessary testing, this will be ready for release.

File: bin/cml/comment/create.js
Comment (casperdcl) [inline bin/cml/comment/create.js:27]: migrated feedback from https://github.com/iterative/cml/pull/1241#discussion_r1024437445 as requested:

I think we need:

- `#123` to auto-infer pr-or-issue
- `mr` alias for `pr`
- `mr#123` alias for `pr#123`
- why do we have `commit#d34db3`? Doesn't `--commit-sha=d34db3` already cover this?
- full URL (e.g. if they want to target another PR returned by `cml pr`)

Comment (tasdomas) [reply bin/cml/comment/create.js:27]: >     * `#123` to auto-infer pr-or-issue
Why? Is there any confusion among users whether they are adding a PR or issue comment?

>     * `mr` alias for `pr`
>     * `mr#123` alias for `pr#123`
Did we have a `--mr` alias for the `--pr` flag?

>     * full URL (e.g. if they want to target another PR returned by `cml pr`)
This could be done, but probably in a follow-up.

> Also, why do we have `commit#d34db3`? Doesn't `--commit-sha=d34db3` already cover this?
As @dacbd mentioned, this is the whole point of `--target` - to get rid of the `--pr`, `--commit-sha` flags (and not introduce `--issue`).

Comment (casperdcl) [reply bin/cml/comment/create.js:27]: ok fine by me to move this to a follow-up issue

**Comment by casperdcl** [comment]
> One of the upsides of this change is that we no longer need `ref: ${{ github.event.pull_request.head.sha }}` to be able to correctly match comments to commits.

so does this address #1204 too?

**Comment by tasdomas** [comment]
> > One of the upsides of this change is that we no longer need `ref: ${{ github.event.pull_request.head.sha }}` to be able to correctly match comments to commits.
> 
> so does this address #1204 too?

`cml check create` seems to resolve the commit sha correctly, so yes, this does address #1204 too

**Review by dacbd — APPROVED** [review]
https://github.com/iterative/cml-playground/actions/runs/3516591620

File: bin/cml/comment/create.js
Comment (casperdcl) [inline bin/cml/comment/create.js:33]:
```suggestion
    conflicts: ['target', 'commitSha'],
    hidden: true
```

File: bin/cml/comment/create.js
Comment (casperdcl) [inline bin/cml/comment/create.js:39]:
```suggestion
    conflicts: ['target', 'pr'],
    hidden: true
```

File: bin/cml/comment/create.js
Comment (casperdcl) [inline bin/cml/comment/create.js:27]:
```suggestion
      'Comment type (e.g. `commit`, `pr`, `issue`, `pr#42`, `issue#1337`),' +
      'default is automatic (`pr` but fallback to `commit`).'
```

File: src/commenttarget.js
Comment (casperdcl) [inline src/commenttarget.js:1]: Should we use this instead (doesn't require escaping, i.e. `--target="pr#1"` => `--target=pr/1`?

```suggestion
const SEPARATOR = '/';
```

Comment (casperdcl) [reply src/commenttarget.js:1]: Yes, because `echo --target pr#1` => `--target pr` since `#` is used for inline shell comments

**Comment by casperdcl** [comment]
Also qq - does this work as expected?

```yaml
on: [issue_comment, pull_request_review_comment]
...
run: cml comment create
```

**Comment by tasdomas** [comment]
> Also qq - does this work as expected?
> 
> ```yaml
> on: [issue_comment, pull_request_review_comment]
> ...
> run: cml comment create
> ```
What is the expected behavior at this point?

Should `cml` automatically create issue comments when running in a `issue_comment` context? It currently does not do that.

The second trigger (`pull_request_review_comment`)  will be handled correctly - the created comment will be a pull request comment.

File: src/commenttarget.js
Comment (tasdomas) [reply src/commenttarget.js:1]: The change is up for review [here](https://github.com/iterative/cml/pull/1279)

**Comment by casperdcl** [comment]
ok [this](https://github.com/iterative/cml/pull/1228#issuecomment-1342815784) isn't a blocker for me - happy to move to a follow-up, but yes I'd expect `on: [issue_comment]` to create comments in the issue by default.

File: src/commenttarget.js
Comment (casperdcl) [inline src/commenttarget.js:10]:
```suggestion
      `Deprecation warning: use --target="commit${SEPARATOR}<sha>" instead of --commit-sha=<sha>`
```

or can we introduce a new proper warning class for these kinds of messages (similar to Python `DeprecationWarning`)?

File: src/commenttarget.js
Comment (casperdcl) [inline src/commenttarget.js:15]:
```suggestion
    drv.warn('Deprecation warning: use --target=pr instead of --pr');
```

File: src/commenttarget.js
Comment (casperdcl) [inline src/commenttarget.js:47]:
```suggestion
    throw new Error(`comment --target="${commentTarget}" could not be parsed`);
```

**Review by casperdcl — APPROVED** [review]
lgtm

File: src/commenttarget.js
Comment (tasdomas) [reply src/commenttarget.js:47]: Warnings I get, but should we capitalize errors too?

File: src/commenttarget.js
Comment (tasdomas) [reply src/commenttarget.js]: Fixed in https://github.com/iterative/cml/pull/1228/commits/dcb37b323fc75872563c163f260bdaf534e76747

File: src/commenttarget.js
Comment (tasdomas) [reply src/commenttarget.js]: Fixed in https://github.com/iterative/cml/pull/1228/commits/dcb37b323fc75872563c163f260bdaf534e76747

**Comment by casperdcl** [comment]
Are there any blockers for this? Too scary for anyone to click "merge"?

**Comment by tasdomas** [comment]
And I can't merge this - need an approval

**Pull request merged**
